Namespace deletion uses discovery + dynamic client

derekwaynecarr opened the pull request:
Fixes: https://github.com/kubernetes/kubernetes/issues/14765
Fixes: https://github.com/kubernetes/kubernetes/issues/21701

The namespace controller:
- takes a list of group version resources supported based on the discovery client
- during deletion, uses dynamic client to remove those resources
  - does DELETE_COLLECTION if supported, otherwise falls back to LIST/DELETE by name.
  - verifies that no items exist for each collection for each resource.

This means developers no longer need to remember to clean-up their resource as part of namespace lifecycle.

derekwaynecarr commented:
see: https://github.com/kubernetes/kubernetes/pull/20351#issuecomment-185333466

derekwaynecarr commented:
@lavalamp @thockin @bgrant0607 - this PR updates the namespace controller to use the discovery client to find the list of supported resources, and the dynamic client to then delete those resources.  this should eliminate developer errors where developers had forgot to do this manually.  assuming this is stable, it may be worth our while to have it in 1.2.

/cc @smarterclayton @kubernetes/rh-cluster-infra

derekwaynecarr commented:
@liggitt - can you take a look as well?

lavalamp commented:
@krousey Can you take a look?

krousey commented:
@lavalamp sure. Is this 1.2 critical?

lavalamp commented:
It's 1.2 nice-to-have.

derekwaynecarr commented:
Looks like my last test broke a unit test, investigating.

@lavalamp @krousey I agree its a nice-to-have for 1.2.  I always worry that people forget to handle deletion issues as things move around API groups, and this just eliminates that worry from the project.

derekwaynecarr commented:
Recording here so I do not forget:

To support re-using the same controller to delete downstream OpenShift specific resources or in federated API server use cases, I will need to let a namespace controller have a finalizer token that it is associated with passed in during creation.

krousey commented on pkg/client/typed/dynamic/client_pool.go:
```diff
@@ -0,0 +1,91 @@
+	if len(groupVersion.Group) == 0 {
+		return "/api"
+	}
+	return "/apis"
```
I know it's not, but shouldn't this be in the discovery into?

derekwaynecarr replied:
opened issue: https://github.com/kubernetes/kubernetes/issues/21916

krousey commented on pkg/client/typed/dynamic/client_pool.go:
```diff
@@ -0,0 +1,91 @@
+}
+
+// ensure that clientPoolImpl conforms to Factory
+var _ ClientPool = &clientPoolImpl{}
```
This isn't needed since you're returning this type into the interface on line 59.

krousey commented on pkg/client/typed/dynamic/client_pool.go:
```diff
@@ -0,0 +1,91 @@
+
+// Package dynamic provides a client interface to arbitrary Kubernetes
+// APIs that exposes common high level operations and exposes common
+// metadata.
```
Duplicate package documentation.

krousey commented on pkg/client/typed/dynamic/client_pool.go:
```diff
@@ -0,0 +1,91 @@
+	// do we have a client already configured?
+	if existingClient, found := c.clients[groupVersion]; found {
+		return existingClient, nil
+	}
```
This check should come first, before the config manipulation.

krousey commented on pkg/controller/namespace/namespace_controller_utils.go:
```diff
@@ -38,6 +43,59 @@ func (e *contentRemainingError) Error() string {
+// operationSupportedCache is a simple cache to remember if an operation is supported for a resource.
+type operationSupportedCache struct {
+	store cache.Store
+}
```
Maybe I'm missing something, but why do we need a full blown cache here? This doesn't seem to need any of the functionality there. Could this be replaced with:

```go
type opKey struct {
	op  operation
	gvr unversioned.GroupVersionResource
}

type operationSupportedCache map[opKey]bool

...
supported, exists := c[opKey{op: op, gvr: gvr}]
```

?

derekwaynecarr replied:
i think i was unncessarily worried about locking late at night when i did this, when i don't think that is a major concern...  i will look to update.

derekwaynecarr commented:
We agreed to punt on re-polling discovery client for resources for kube 1.2 in api machinery sig.

derekwaynecarr commented:
@krousey - I simplified the operation supported cache, and made the other updates requested.

@smarterclayton @liggitt - I updated this to take a finalizer token as input when creating namespace controller, so if/when we use this controller in downstream OpenShift, we can pass in the "openshift.io/origin" token and hopefully remove our custom controller downstream.

derekwaynecarr commented:
@k8s-bot test this ISSUE #21798

derekwaynecarr commented:
Rebased to hopefully pick up the flake fix in https://github.com/kubernetes/kubernetes/pull/21857

derekwaynecarr commented:
Looks like a different flake this time.  Will investigate in morning.

On Wednesday, February 24, 2016, Kubernetes Bot <notifications@github.com>
wrote:

> GCE e2e build/test *failed* for commit a577ea3
> <https://github.com/kubernetes/kubernetes/commit/a577ea3c1de1ab89eaa3125af7bc68dbc6dc4c00>
> .
>
>    - Build Log
>    <http://pr-test.k8s.io/21400/kubernetes-pull-build-test-e2e-gce/30312/build-log.txt>
>    - Test Artifacts
>    <https://console.developers.google.com/storage/browser/kubernetes-jenkins/pr-logs/pull/21400/kubernetes-pull-build-test-e2e-gce/30312/artifacts/>
>    - Internal Jenkins Results
>    <http://goto.google.com/prkubekins/job/kubernetes-pull-build-test-e2e-gce//30312>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/21400#issuecomment-188535040>
> .
>

ncdc commented:
```
Failure [342.411 seconds]
Kubectl client
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:1082
  Kubectl describe
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:653
    should check if kubectl describe prints relevant information for rc and pods [Conformance] [It]
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:652

    Expected error:
        <*errors.errorString | 0xc2084fddb0>: {
            s: "gave up waiting for pod 'redis-master-70oc8' to be 'running' after 5m0s",
        }
        gave up waiting for pod 'redis-master-70oc8' to be 'running' after 5m0s
    not to have occurred

    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/examples.go:483
```

krousey commented on pkg/controller/namespace/namespace_controller_utils.go:
```diff
@@ -103,99 +130,145 @@ func finalizeNamespaceFunc(kubeClient clientset.Interface, namespace *api.Namesp
+			gvr: groupVersionResource,
+		}
+		canDeleteCollection, canDeleteCollectionKnown := operationSupportedCache[canDeleteCollectionKey]
+		canList, canListKnown := operationSupportedCache[canListKey]
```
You could probably simplify this if you changed the semantics of the cache from "operation supported" to "operation not supported". Then to check if delete might succeed, just

```go
if !cache[deleteKey] {
	...
}
```

By inverting the semantic, you wouldn't need to deal with map presence detection.

krousey commented:
One minor nit, otherwise it looks good to me.

derekwaynecarr commented:
@krousey - updated cache one last time.

see: https://github.com/derekwaynecarr/kubernetes/commit/ddea93143d6c003354a10d4f47fb9362b5c10700#diff-f86a8dd1c06d1f9578423999f8094cc7R181

I think this is good to go...

derekwaynecarr commented:
@k8s-bot test this issue #22011

derekwaynecarr commented:
@lavalamp - I think this should be ready for a final review.

derekwaynecarr commented:
@k8s-bot test this issue #22187

lavalamp commented on pkg/client/typed/dynamic/client_pool.go:
```diff
@@ -0,0 +1,85 @@
+)
+
+// ClientPool manages a pool of dynamic clients.
+type ClientPool interface {
```
Awesome. Later we can make kubectl's builder use this, perhaps.

lavalamp commented on pkg/controller/namespace/namespace_controller_utils.go:
```diff
@@ -264,125 +327,76 @@ func syncNamespace(kubeClient clientset.Interface, versions *unversioned.APIVers
+			if !apiResource.Namespaced {
+				continue
+			}
+			if strings.Contains(apiResource.Name, "/") {
```
Is this looking for subresources?

derekwaynecarr replied:
this was intended to filter out subresources...

lavalamp replied:
Did we put subresources in that list? I thought that was still a todo...

ncdc replied:
Someone must have implemented it - subs are in

lavalamp commented on pkg/controller/namespace/namespace_controller_utils.go:
```diff
@@ -264,125 +327,76 @@ func syncNamespace(kubeClient clientset.Interface, versions *unversioned.APIVers
 	}
 	estimate := int64(0)
 	for i := range items.Items {
+		// filter out terminal pods
```
Is this new in this PR? Is it correct?

derekwaynecarr replied:
this is new in this PR and was intended to fix:
https://github.com/kubernetes/kubernetes/issues/21701

If a pod is in a terminal state, there is no needed to unnecessarily wait for graceful deletion of that pod because its already terminated.

lavalamp commented on pkg/controller/namespace/namespace_controller_utils.go:
```diff
@@ -103,99 +130,135 @@ func finalizeNamespaceFunc(kubeClient clientset.Interface, namespace *api.Namesp
-			}
+		var deleteOptions *v1.DeleteOptions
+		if supportsGrace {
+			deleteOptions = &v1.DeleteOptions{GracePeriodSeconds: &groupVersionResourceEstimate}
```
Don't pods have a default baked in? Do you really need to override the default?

lavalamp commented on pkg/controller/namespace/namespace_controller_utils.go:
```diff
@@ -38,6 +42,23 @@ func (e *contentRemainingError) Error() string {
+	gvr unversioned.GroupVersionResource
+}
+
+// operationNotSupportedCache is a simple cache to remember if an operation is not supported for a resource.
```
Document that "true" == "not supported"?

lavalamp commented on pkg/controller/namespace/namespace_controller_utils.go:
```diff
@@ -103,99 +130,135 @@ func finalizeNamespaceFunc(kubeClient clientset.Interface, namespace *api.Namesp
+		groupVersionResourceEstimate, supportsGrace, err := estimateGracefulTermination(kubeClient, groupVersionResource, namespace, namespaceDeletedAt)
 		if err != nil {
+			glog.V(4).Infof("namespace controller - deleteAllContent - unable to estimate - namespace: %s, resource: %v, err: %v", namespace, groupVersionResource, err)
 			return estimate, err
```
I don't think these returns are appropriate any more, since they cause the rest of the resources to be skipped?

lavalamp commented on pkg/controller/namespace/namespace_controller_utils.go:
```diff
@@ -103,99 +130,135 @@ func finalizeNamespaceFunc(kubeClient clientset.Interface, namespace *api.Namesp
+	for _, groupVersionResource := range groupVersionResources {
+		glog.V(4).Infof("namespace controller - deleteAllContent - deleting resource - namespace: %s, resource: %v", namespace, groupVersionResource)
+		// estimate how long it will take for the resource to be deleted (needed for objects that support graceful delete)
+		groupVersionResourceEstimate, supportsGrace, err := estimateGracefulTermination(kubeClient, groupVersionResource, namespace, namespaceDeletedAt)
```
I don't know if I want to block this PR or not, but this function would be a lot more understandable (and bug free?) if:

* Move the body of this for loop to another function
* Move every cache-check/operation section to its own function
* Add a function to the cache like "isSupported" instead of the !cache[op] pattern.

In particular, I kind of think that many of the `return`s in this function body are actually bugs?

derekwaynecarr replied:
I think this all makes total sense.  I will update tomorrow and poke the PR.

On Monday, February 29, 2016, Daniel Smith <notifications@github.com> wrote:

> In pkg/controller/namespace/namespace_controller_utils.go
> <https://github.com/kubernetes/kubernetes/pull/21400#discussion_r54480441>
> :
>
> > +// If estimate > 0, not all resources are guaranteed to be gone.
> > +func deleteAllContent(
> > +	kubeClient clientset.Interface,
> > +	clientPool dynamic.ClientPool,
> > +	opCache operationNotSupportedCache,
> > +	groupVersionResources []unversioned.GroupVersionResource,
> > +	namespace string,
> > +	namespaceDeletedAt unversioned.Time,
> > +) (int64, error) {
> > +	estimate := int64(0)
> > +	glog.V(4).Infof("namespace controller - deleteAllContent - namespace: %s, resources: %v", namespace, groupVersionResources)
> > +	// iterate over each group version, and attempt to delete all of its resources
> > +	for _, groupVersionResource := range groupVersionResources {
> > +		glog.V(4).Infof("namespace controller - deleteAllContent - deleting resource - namespace: %s, resource: %v", namespace, groupVersionResource)
> > +		// estimate how long it will take for the resource to be deleted (needed for objects that support graceful delete)
> > +		groupVersionResourceEstimate, supportsGrace, err := estimateGracefulTermination(kubeClient, groupVersionResource, namespace, namespaceDeletedAt)
>
> I don't know if I want to block this PR or not, but this function would be
> a lot more understandable (and bug free?) if:
>
>    - Move the body of this for loop to another function
>    - Move every cache-check/operation section to its own function
>    - Add a function to the cache like "isSupported" instead of the
>    !cache[op] pattern.
>
> In particular, I kind of think that many of the returns in this function
> body are actually bugs?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/21400/files#r54480441>.
>

lavalamp commented:
I finally got a chance to go through this. @derekwaynecarr sorry it took me so long.

derekwaynecarr committed:
Add ability to manage a pool of dynamic clients

derekwaynecarr commented:
@lavalamp - rebased, and updated per code review requests.  PTAL

derekwaynecarr committed:
Move namespace controller to use discovery and dynamic client

derekwaynecarr committed:
Add e2e tests to verify more resources are deleted

derekwaynecarr committed:
Fix incorrect logging string that could cause confusion

eparis commented:
@k8s-bot test this, issue #22435

lavalamp commented on pkg/controller/namespace/namespace_controller_utils.go:
```diff
@@ -103,99 +137,207 @@ func finalizeNamespaceFunc(kubeClient clientset.Interface, namespace *api.Namesp
+	// this is strange, but we need to special case for both MethodNotSupported and NotFound errors
+	// TODO: https://github.com/kubernetes/kubernetes/issues/22413
+	// we have a resource returned in the discovery API that supports no top-level verbs:
+	//  /apis/extensions/v1beta1/namespaces/default/replicationcontrollers
```
So, rather than this special case here, you may want to special case just this and keep it out of the resource list entirely. That path only exists to mount a subresource. We won't ever make anything like it again.

lavalamp commented on pkg/controller/namespace/namespace_controller_utils.go:
```diff
@@ -103,99 +137,207 @@ func finalizeNamespaceFunc(kubeClient clientset.Interface, namespace *api.Namesp
+func listCollection(
+	dynamicClient *dynamic.Client,
+	opCache operationNotSupportedCache,
+	gvr unversioned.GroupVersionResource,
```
I hate myself for slowing this down even more, but I feel compelled to mention that making an object to store some of these parameters and making some of these functions methods on the object would probably be cleaner. But feel free to ignore this pedantry.

derekwaynecarr replied:
No disagreement, I think in some cases, the code is a reflection of the problems I encountered when doing generic client work, and ideally, I would have never needed this opCache, etc.  The proliferation of arguments came as a response to the iterative process.  A follow-up should be fine I think.

lavalamp commented on pkg/controller/namespace/namespace_controller_utils.go:
```diff
@@ -103,99 +137,207 @@ func finalizeNamespaceFunc(kubeClient clientset.Interface, namespace *api.Namesp
+	groupVersionResources []unversioned.GroupVersionResource,
+	namespace string,
+	namespaceDeletedAt unversioned.Time,
+) (int64, error) {
```
How does an int64 map to time? (why not time.Duration?)

derekwaynecarr replied:
I kept the type consistent with gracefulTermination which is seconds.

lavalamp commented on pkg/controller/namespace/namespace_controller_utils.go:
```diff
@@ -103,99 +137,207 @@ func finalizeNamespaceFunc(kubeClient clientset.Interface, namespace *api.Namesp
+	for _, gvr := range groupVersionResources {
+		gvrEstimate, err := deleteAllContentForGroupVersionResource(kubeClient, clientPool, opCache, gvr, namespace, namespaceDeletedAt)
 		if err != nil {
 			return estimate, err
```
should this be `return 0, err`?

derekwaynecarr replied:
It will ignore the estimate and only look at the err.

lavalamp commented:
So I have a bunch of nitpicky things that I said and some I didn't say. But if you're confident that this works, I'm willing to hold my nose and LGTM.

derekwaynecarr commented:
@lavalamp - I'm confident it works, as for holding noses, if we are fine holding our nose on `ReplicationControllerDummy` (https://github.com/kubernetes/kubernetes/issues/22413), I think this smells a little better than what we had before and vote to merge for 1.2.

lavalamp commented:
LGTM thanks!
